types/node/async_hooks.d.ts

```diff
@@ -199,7 +199,7 @@ declare module "async_hooks" {
          * stacktrace will not be impacted by this call and the context will be exited.
          */
         // TODO: Apply generic vararg once available
-        run<R>(store: T, callback: (...args: any[]) => R, ...args: any[]): R;
+        run(store: T, callback: (...args: any[]) => void, ...args: any[]): Promise<void>;
```

Comment:
I don't think this is correct. `run` calls `callback` and returns the result from the callback. Therefore I think the current definition is correct.

Reply:
Are you sure about that? The documentation seems to disagree.

Also: If `callback` is invoked asynchronously, how can `run()` return the result of that callback when it hasn't been invoked yet?

Reply:
Yes, the callback is called sync, see https://github.com/nodejs/node/blob/354b6a93bd1d66f1833489d6fe01a2b0e8f6aff9/lib/async_hooks.js#L292-L305

Reply:
Looking at the actual *implementation* I would have to agree. The return value *is* of type `R`, but that means that the documentation for this function is clearly wrong, because it says the result should be awaited:

![image](https://user-images.githubusercontent.com/411338/96840419-b747bf00-144a-11eb-87af-42976b1a08ee.png)

I'll revert this change.

Reply:
Sample:
```
const { AsyncLocalStorage} = require("async_hooks");
const als = new AsyncLocalStorage();
console.log(typeof als.run({}, () => 15));
console.log(typeof als.run({}, () => "15"));
```
prints
```
number
string
```

Reply:
Updated sample:
```
async function foo() {
    console.log(typeof als.run({}, () => 15));
    console.log(typeof als.run({}, () => "15"));
    console.log(typeof await als.run({}, () => 15));
    console.log(typeof await als.run({}, () => "15"));
}
foo();

prints
```
number
string
number
string
```
`await` magically converts the sync return value into a `Promise` and awaits it.
Usually you want do it that way instead the called function is a function returning a `Promise`.

I think the node docs are a little bit misleading as they may be interpreted that users have to `await` the `als.run` call. But that is not true.
The intention of this doc is to explain that you have to take care with using `AsyncLocalStorage` within async functions. Maybe I will file a PR against node docs if I have a good idea how to describe this better.